### ISSUE TITLE
to_device not working

### ISSUE THREAD

**Issue by SupaMaggie70Incorporated** [tracel-ai/burn#1890]
**Describe the bug**
When a `Module` is created on a GPU device(e.g. Cuda on candle or tch, DiscreteGpu on wgpu), training is just fine. However, when a model is created on cpu and sent to the gpu using Module::to_device(), it is unable to train, though it can evaluate and give losses just fine. The "training" completes way faster, except for the first 1-3 iterations likely caused by first use of functions, but the model appears to not be updated at all.

**To Reproduce**
Create a `Module` on the cpu, send it to the gpu using `to_device()`, and attempt to train it.

**Expected behavior**
One would expect the model to train

**Desktop (please complete the following information):**
 - OS: Ubuntu 2022.04
 - Backends: Candle, tch, wgpu, probably all due to Autodiffbackend
 - Burn version 13.2, most recent at time of this bug report

**Smartphone (please complete the following information):**

**Additional context**
This is probably an issue with AutodiffBackend, as it is shared across backends, and prevents gradients/training but doesn't prevent evaluation and loss.

**Comment by nathanielsimard** [comment]
Experimenting with this a bit, I don't think there is a bug, but maybe a misuse of `Module::to_device`. You can't initialize a module on a device, switch it to another one using `to_device`, and expect the gradients to be kept for the module on the new device. The new module, on the GPU, is considered an in-between node where the gradients are consumed during backprop, and are only useful to compute the gradients for the first module (on the CPU).

What you are probably searching for is the method `Module::fork` where there is no link between the parent module and the new moved module, which can be optimized. Just to illustrate:

```rust
#[test]
fn module_to_device() {
    type B = Autodiff<Wgpu>;
    let device_gpu = WgpuDevice::default();
    let device_cpu = WgpuDevice::Cpu;

    let initializer = Initializer::Ones;

    let tensor_cpu: Param<Tensor<B, 1>> = initializer.init_with([32], None, None, &device_cpu);
    let tensor_gpu = tensor_cpu.clone().to_device(&device_gpu);

    let loss = tensor_gpu.val().tanh().sum();

    let grads = loss.backward();

    let grad_gpu = grads.get(&tensor_gpu.val().into_primitive());
    let grad_cpu = grads.get(&tensor_cpu.val().into_primitive());

    assert!(grad_cpu.is_some(), "Passes");
    assert!(grad_gpu.is_some(), "Fails");
}
```

But it's the oposite when you use fork:

```rust
#[test]
fn module_to_device() {
    type B = Autodiff<Wgpu>;
    let device_gpu = WgpuDevice::default();
    let device_cpu = WgpuDevice::Cpu;

    let initializer = Initializer::Ones;

    let tensor_cpu: Param<Tensor<B, 1>> = initializer.init_with([32], None, None, &device_cpu);
    let tensor_gpu = tensor_cpu.clone().fork(&device_gpu);

    let loss = tensor_gpu.val().tanh().sum();

    let grads = loss.backward();

    let grad_gpu = grads.get(&tensor_gpu.val().into_primitive());
    let grad_cpu = grads.get(&tensor_cpu.val().into_primitive());

    assert!(grad_gpu.is_some(), "Passes");
    assert!(grad_cpu.is_some(), "Fails");
}
```

The documentation on `Module::to_device` is warning users of that behavior:

```
The device operations will be registered in the autodiff graph. Therefore, be sure to call
backward only one time even if you have the same module on multiple devices. If you want to
call backward multiple times, look into using fork.
```

This isn't very clear, and not 100% accurate, so we should update it.

**Comment by SupaMaggie70Incorporated** [comment]
@nathanielsimard 
I had read this but didn't realize you had to compute the gradients for the original device network. After all, I wasn't doing much with the cpu, all the gradients were to be generated and used on the gpu. I will test your proposal when I get back home, though it will most likely solve my problem. I would just point out that in PyTorch at least, Tensor.to doesn’t have this behavior, and it might make more sense to either clarify that or rename some methods to be less confusing. Of course I’m an ml noob, and this would’ve been pretty straightforward if the docs had just been clarified as you suggested doing.
